cloudinit/cmd/devel/logs.py

```diff
@@ -196,131 +171,207 @@ def _stream_command_output_to_file(
+    ]
+    yield from (
+        path
+        for path in etc_cloud_dir.glob("**/*")
+        if path.name not in ignore and path.parent not in ignore
+    )
+
+
+def _get_var_lib_cloud(cloud_dir: pathlib.Path) -> Iterator[pathlib.Path]:
+    """Get paths for files in /var/lib/cloud.
+
+    Skip user-provided scripts, semaphores, and old instances.
+    """
+    return itertools.chain(
+        cloud_dir.glob("data/*"),
+        cloud_dir.glob("handlers/*"),
+        cloud_dir.glob("seed/*"),
+        (p for p in cloud_dir.glob("instance/*") if p.is_file()),
+        cloud_dir.glob("instance/handlers"),
+    )
+
+
+def _get_run_dir(run_dir: pathlib.Path) -> Iterator[pathlib.Path]:
+    """Get all paths under /run/cloud-init except for hook-hotplug-cmd.
+
+    Note that this only globs the top-level directory as there are currently
+    no relevant files within subdirectories.
+    """
+    return (p for p in run_dir.glob("*") if p.name != "hook-hotplug-cmd")
+
+
+def _collect_logs_into_tmp_dir(
+    log_dir: pathlib.Path,
+    log_cfg: Dict[str, Any],
+    run_dir: pathlib.Path,
+    cloud_dir: pathlib.Path,
+    include_sensitive: bool,
+) -> None:
+    """Collect all cloud-init logs into the provided directory."""
+    _collect_version_info(log_dir)
+    _collect_system_logs(log_dir, include_sensitive)
+    _collect_installer_logs(log_dir, include_sensitive)
+
+    for logfile in _get_cloudinit_logs(log_cfg):
+        # Even though log files are root read-only, the logs tarball
+        # would be useless without them and we've been careful to not
+        # include sensitive data in them.
+        _collect_file(
+            logfile,
+            log_dir / pathlib.Path(logfile).parent.relative_to("/"),
+            True,
         )
-        return 1
+    for logfile in itertools.chain(
+        _get_etc_cloud(),
+        _get_var_lib_cloud(cloud_dir=cloud_dir),
+        _get_run_dir(run_dir=run_dir),
+    ):
+        _collect_file(
+            logfile,
+            log_dir / pathlib.Path(logfile).parent.relative_to("/"),
+            include_sensitive,
+        )
+
+
+def collect_logs(
+    tarfile: str,
+    log_cfg: Dict[str, Any],
+    run_dir: pathlib.Path = pathlib.Path("/run/cloud-init"),
+    cloud_dir: pathlib.Path = pathlib.Path("/var/lib/cloud"),
+    include_sensitive: bool = True,
+) -> None:
+    """Collect all cloud-init logs and tar them up into the provided tarfile.
 
+    :param tarfile: The path of the tar-gzipped file to create.
+    :param log_cfg: The cloud-init base configuration containing logging cfg.
+    :param run_dir: The path to the cloud-init run directory.
+    :param cloud_dir: The path to the cloud-init cloud directory.
+    :param include_sensitive: Boolean, true means include sensitive data.
+    """
     tarfile = os.path.abspath(tarfile)
-    log_dir = (
+    dir_name = (
         datetime.now(timezone.utc).date().strftime("cloud-init-logs-%Y-%m-%d")
     )
     with tempdir(dir="/tmp") as tmp_dir:
```

Comment:
This might look scary from a permissions perspective, especially since we're now requiring root and then writing to `/tmp` (which has open permissions), however this uses `tempfile.mkdtemp()` under the covers so it is actually fine. Perhaps we could make alleviate concerns by changing this to `/run/cloud-init/`?